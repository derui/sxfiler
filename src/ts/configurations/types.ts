import { ObjectEnum } from "@/utils";

export type ItemKey<T> = readonly [string, string, string];
export type SectionKey = readonly [string, string];
export type CategoryKey = readonly [string];

export type Descriptor<T> = {
  readonly key: T;
  readonly displayName: string;
  readonly description: string;
};

export type CategoryDescriptor = Descriptor<CategoryKey>;

export type SectionDescriptor = Descriptor<SectionKey>;

export type ItemDescriptor = Descriptor<ItemKey<any>>;

export type SelectOption = {
  value: any;
  toString(): string;
  display: string;
};

export const ItemElementType = {
  TEXT: "TEXT",
  NUMBER: "NUMBER",
  BOOLEAN: "BOOLEAN",
  SELECT_ONE: "SELECT_ONE",
} as const;
export type ItemElementType = ObjectEnum<typeof ItemElementType>;

export type ItemKind<K extends ItemElementType> = {
  kind: K;
};

export type TextItemType = ItemKind<"TEXT"> & {
  defaultValue?: string;
};
export type NumberItemType = ItemKind<"NUMBER"> & {
  defaultValue?: number;
};
export type BooleanItemType = ItemKind<"BOOLEAN"> & {
  defaultValue: boolean;
};
export type SelectOneItemType = ItemKind<"SELECT_ONE"> & {
  options: SelectOption[];
  defaultValue: string;
};

export type ItemType = TextItemType | NumberItemType | BooleanItemType | SelectOneItemType;

export type Item = ItemDescriptor & {
  readonly type: ItemType;
};

export type Section = SectionDescriptor & {
  readonly items: Item[];
};

export type Category = CategoryDescriptor & {
  readonly sections: Section[];
};
