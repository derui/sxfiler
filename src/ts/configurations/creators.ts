import { Section, Category, Item, ItemType, ItemKey, SectionDescriptor, CategoryDescriptor } from "./types";

export const createCategory = function (
  args: Omit<CategoryDescriptor, "key"> & { key: readonly [string]; sections: Section[] }
): Category {
  const { key, ...rest } = args;
  return { key: { key: key }, ...rest };
};

export const createSection = function (
  args: Omit<SectionDescriptor, "key"> & { key: readonly [string, string]; items: Item[] }
): Section {
  const { key, ...rest } = args;
  return { key: { key: key }, ...rest };
};

export const toItemKey = function <T = any>(key: [string, string, string]): ItemKey<T> {
  return {
    type_: undefined,
    key,
  };
};

export const createItem = function (args: {
  key: ItemKey<any>;
  displayName: string;
  description: string;
  type: ItemType;
}): Item {
  const { key, ...rest } = args;
  return {
    key,
    displayName: rest.displayName,
    description: rest.description,
    type: rest.type,
  };
};
