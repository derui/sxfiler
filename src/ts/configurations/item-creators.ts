import { ItemElementType, ItemType, SelectOption } from "./types";

export const createText = (value?: string): ItemType => ({ kind: ItemElementType.TEXT, defaultValue: value });

export const createNumber = (value?: number): ItemType => ({ kind: ItemElementType.NUMBER, defaultValue: value });

export const createBoolean = (value: boolean): ItemType => ({ kind: ItemElementType.BOOLEAN, defaultValue: value });

export const createSelectOne = (options: SelectOption[], value: string): ItemType => ({
  kind: ItemElementType.SELECT_ONE,
  options,
  defaultValue: value,
});
