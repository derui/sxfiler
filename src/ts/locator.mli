open Locator_abbrevs

include module type of struct
  include Locator_intf
end

(** A signature for store instance to make locator instance. *)
module type Store_instance = sig
  val this : S.App.Store.t
end

val make_store : unit -> S.App.Store.t
(** [make_store ()] gets initial store of application  *)

module Make
    (Client : C.Rpc_client.S)
    (Context : Sxfiler_renderer_core.Context.Instance)
    (Store : Store_instance) : S
