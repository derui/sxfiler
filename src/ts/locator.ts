import { Context as ReactContext, createContext } from "react";

import { ContextLike } from "./context";

export type Locator = {
  context?: ContextLike;
};

let locator: Locator = {};

/**
 * Set new locator to global context
 * @param newLocator
 */
export function setLocator(newLocator: Locator): void {
  Object.assign(locator, newLocator);
}

const LocatorContext: ReactContext<Locator> = createContext(locator);

export default LocatorContext;
