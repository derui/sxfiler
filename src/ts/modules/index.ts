//#IMPORT INDICATOR
import * as Configuration from "./configuration";
import * as Keymap from "./keymap";
import * as Filer from "./filer";
import * as Completer from "./completer";
import * as Decision from "./decision";
// prettier-ignore
import { combineReducers } from "redux";

// export combined types
// prettier-ignore
export type Actions =
  | Configuration.Actions
  | Keymap.Actions
  | Filer.Actions
  | Completer.Actions
  | Decision.Actions
  ;

// export combined reducer
// prettier-ignore
export const reducer = combineReducers({
  configuration: Configuration.reducer,
  keymap: Keymap.reducer,
  filer: Filer.reducer,
  completer: Completer.reducer,
  decision: Decision.reducer,
})

export const emptyState = {
  configuration: Configuration.emptyState,
  keymap: Keymap.emptyState,
  filer: Filer.emptyState,
  completer: Completer.emptyState,
  decision: Decision.emptyState,
};

// export state type
export type State = ReturnType<typeof reducer>;
