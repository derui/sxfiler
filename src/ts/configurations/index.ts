import { SortType, SortTypeMap } from "@/generated/types_pb";
import { Category, Item, Section, ItemKey, SectionKey, CategoryKey } from "./types";
import { createCategory, createSection, createItem } from "./creators";
import * as ItemCreators from "./item-creators";

const itemKeyFrom = <T>(section: SectionKey, key: string): ItemKey<T> => [section[0], section[1], key];
const categoryKeyOf = (key: ItemKey<any>): CategoryKey => [key[0]];
const sectionKeyOf = (key: ItemKey<any>): SectionKey => [key[0], key[1]];
const sectionKeyFrom = (category: CategoryKey, key: string): SectionKey => [category[0], key];
export const qualified = (key: readonly string[]): string => key.join(".");

export const findItemBy = (key: ItemKey<any>, category: Category): Item | undefined => {
  const categoryKey = categoryKeyOf(key);
  const sectionKey = sectionKeyOf(key);
  if (qualified(category.key) !== qualified(categoryKey)) {
    return;
  }
  const section = category.sections.find((v) => v.key === sectionKey);
  if (!section) {
    return;
  }

  return section.items.find((v) => qualified(v.key) === qualified(key));
};

export const findSectionBy = (key: SectionKey, category: Category): Section | undefined => {
  return category.sections.find((v) => qualified(v.key) === qualified(key));
};

export const findValueBy = (configuration: { [p: string]: any }, item: Item): any => {
  const value = configuration[qualified(item.key)];
  if (value === undefined || value === null) {
    return;
  }

  return value;
};

export const categories = {
  general: ["general"],
} as const;

export const sections = {
  behaviors: sectionKeyFrom(categories.general, "behaviors"),
  theme: sectionKeyFrom(categories.general, "theme"),
} as const;

export const itemKeys = {
  general: {
    behaviors: {
      confirmationWhenDelete: itemKeyFrom<string>(sections.behaviors, "confirmation_when_delete"),
      maxNumberOfHistories: itemKeyFrom<number>(sections.behaviors, "max_number_of_histories"),
      defaultSortOrder: itemKeyFrom<SortTypeMap>(sections.behaviors, "default_sort_order"),
    },
    theme: {
      currentTheme: itemKeyFrom<string>(sections.theme, "current_theme"),
    },
  },
} as const;

export namespace Configurations {
  const general = createCategory({
    key: categories.general,
    displayName: "General",
    description: "General configurations",
    sections: [
      createSection({
        key: sections.behaviors,
        displayName: "Behaviors",
        description: "Behaviors",
        items: [
          createItem({
            key: itemKeys.general.behaviors.confirmationWhenDelete,
            displayName: "Confirmation when delete",
            description: "Need user confirmation when delete something",
            type: ItemCreators.createBoolean(true),
          }),
          createItem({
            key: itemKeys.general.behaviors.maxNumberOfHistories,
            displayName: "Maximum number of histories",
            description:
              "History is the history of moved directory in a pane. This setting defines maximum number of histories in a pane",
            type: ItemCreators.createNumber(100),
          }),
          createItem({
            key: itemKeys.general.behaviors.defaultSortOrder,
            displayName: "Default sort order of Pane",
            description: "This setting defines sort order of filer when open this application",
            type: ItemCreators.createSelectOne(
              [
                {
                  display: "Date",
                  toString() {
                    return `${SortType.DATE}`;
                  },
                  value: SortType.DATE,
                },
                {
                  display: "Name",
                  toString() {
                    return `${SortType.NAME}`;
                  },
                  value: SortType.NAME,
                },
                {
                  display: "Size",
                  toString() {
                    return `${SortType.SIZE}`;
                  },
                  value: SortType.SIZE,
                },
              ],
              `${SortType.NAME}`
            ),
          }),
        ],
      }),
      createSection({
        key: sections.theme,
        displayName: "Theme",
        description: "Theme",
        items: [
          createItem({
            key: itemKeys.general.theme.currentTheme,
            displayName: "Current theme",
            description: "Select current theme",
            type: ItemCreators.createText("default"),
          }),
        ],
      }),
    ],
  });

  export const definitions = {
    general,
  } as const;
}
