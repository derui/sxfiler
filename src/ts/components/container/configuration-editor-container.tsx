import { h, FunctionalComponent } from "preact";
import { Component as SectionEditor } from "@/components/project/configuration-section";
import { Component as Navigator } from "@/components/project/configuration-navigator";
import { qualified, Configurations, findSectionBy, findValueBy } from "@/configurations";
import * as C from "@/modules";
import { Item, Section, Category } from "@/configurations/types";
import { Locator } from "@/locator";
import { descriptors } from "@/commands/internal";
import { isClosed } from "@/modules/configuration/selectors";

export type Props = {
  state: C.State;
  locator: Locator;
};

const handleSectionSelected = (locator: Locator, state: C.State) => (_: Category, section: Section) => {
  const command = locator.commandResolver?.resolveBy(descriptors.configurationSelectSection);
  if (!command) {
    return;
  }

  locator.commandExecutor?.execute(command, state, { section });
};

const handleUpdated = (locator: Locator, state: C.State) => (item: Item, value: any) => {
  const command = locator.commandResolver?.resolveBy(descriptors.configurationUpdate);
  if (!command) {
    return;
  }

  locator.commandExecutor?.execute(command, state, { itemKey: item.key, value });
};

export const Component: FunctionalComponent<Props> = ({ state, locator }) => {
  const categories = [Configurations.definitions.general];
  const itemValueMap: { [p: string]: any } = {};
  let section;
  const selectedSection = state.configuration.selectedSection;
  if (selectedSection) {
    section = findSectionBy(selectedSection, categories[0]);
    if (section) {
      section.items.reduce((accum, item) => {
        accum[qualified(item.key)] = findValueBy(state.configuration.configuration, item);
        return accum;
      }, itemValueMap);
    }
  }

  const closed = isClosed(state.configuration);

  return (
    <div class="configuration-editor" aria-hidden={closed} data-testid="configuration-editor">
      <Navigator categories={categories} onSectionSelected={handleSectionSelected(locator, state)} />
      {section && (
        <SectionEditor section={section} onUpdated={handleUpdated(locator, state)} itemValueMap={itemValueMap} />
      )}
    </div>
  );
};
