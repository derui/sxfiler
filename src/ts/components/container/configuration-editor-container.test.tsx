import { h } from "preact";
import { render as renderToString } from "preact-render-to-string";

import { emptyState } from "@/modules";
import { Component } from "./configuration-editor-container";
import { Locator } from "@/locator";
import { cleanup, render, fireEvent, act } from "@testing-library/preact";
import { Configurations } from "@/configurations";

describe("Container", () => {
  describe("Configuration Editor Container", () => {
    const executor = jest.fn();
    const resolveBy = jest.fn();
    let locator: Locator | null;

    function makeMockedLocator(): Locator {
      return {
        commandExecutor: {
          execute: executor,
        },
        commandResolver: {
          register: jest.fn(),
          resolveById: jest.fn(),
          resolveBy: resolveBy,
        },
      };
    }
    beforeEach(() => {
      locator = makeMockedLocator();
    });

    afterEach(() => {
      locator = null;
      jest.clearAllMocks();
      cleanup();
    });

    it("render correctly when context not initialized and empty state", () => {
      const locator = {};
      const state = emptyState;

      const wrapper = renderToString(<Component state={state} locator={locator} />);

      expect(wrapper).toMatchSnapshot();
    });

    it("call command if section clicked", async () => {
      const state = emptyState;

      const query = render(<Component state={state} locator={locator!!} />);

      const section = await query.findAllByTestId("configuration-navigator-section");
      const command = jest.fn();
      resolveBy.mockImplementation(() => command);

      act(() => {
        fireEvent.click(section[0]);
      });

      expect(executor).toBeCalledWith(command, state, { section: Configurations.definitions.general.sections[0] });
    });
  });
});
