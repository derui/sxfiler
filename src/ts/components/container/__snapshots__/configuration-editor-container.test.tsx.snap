// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Configuration Editor Container render correctly when context not initialized and empty state 1`] = `
"<div class=\\"configuration-editor\\" aria-hidden=\\"true\\" data-testid=\\"configuration-editor\\"><section class=\\"configuration-navigator\\"><div class=\\"configuration-navigator__category\\" data-testid=\\"configuration-navigator-category\\"><span class=\\"
    configuration-navigator__category-name
    
  \\" data-testid=\\"configuration-navigator-category-name\\">General</span><ul class=\\"configuration-navigator__section-tree\\" data-testid=\\"configuration-navigator-section-tree\\" aria-hidden=\\"true\\"><li class=\\"configuration-navigator__section\\" data-testid=\\"configuration-navigator-section\\">Behaviors</li><li class=\\"configuration-navigator__section\\" data-testid=\\"configuration-navigator-section\\">Theme</li></ul></div></section></div>"
`;
